src/GuiBase/Viewer/Gizmo/Gizmo.cpp

```diff
@@ -68,20 +82,61 @@ bool Gizmo::findPointOnPlane( const Engine::Camera& cam,
     return hasHit;
 }
 
-void Gizmo::addRenderObject( Engine::RenderObject* ro, const std::shared_ptr<Engine::Mesh>& mesh ) {
-    m_meshes.push_back( mesh );
-    m_renderObjects.push_back( m_comp->addRenderObject( ro ) );
+void Gizmo::addRenderObject( Engine::RenderObject* ro ) {
+    m_comp->addRenderObject( ro );
+    m_ros.push_back( ro );
 }
 
-Engine::RenderTechnique* Gizmo::makeRenderTechnique( const std::string& mtlName,
-                                                     bool rtPerVertexColor ) {
-    auto rt       = new Engine::RenderTechnique;
+std::shared_ptr<Engine::RenderTechnique> Gizmo::makeRenderTechnique( int color ) {
+    auto rt       = Core::make_shared<Engine::RenderTechnique>();
     auto plaincfg = Engine::ShaderConfigurationFactory::getConfiguration( "Plain" );
+    auto provider = new Gizmo::UiSelectionControler( s_material[color] );
     rt->setConfiguration( *plaincfg );
-    auto mat              = Core::make_shared<Engine::PlainMaterial>( mtlName );
-    mat->m_perVertexColor = rtPerVertexColor;
-    rt->setParametersProvider( mat );
+    rt->setParametersProvider( std::shared_ptr<Engine::ShaderParameterProvider>( provider ) );
     return rt;
 }
+
+Gizmo::UiSelectionControler* Gizmo::getControler( int ro ) {
+    // TODO : can these casts be replaced by a improved design ?
+    auto c = dynamic_cast<const Gizmo::UiSelectionControler*>(
+        m_ros[ro]->getRenderTechnique()->getParametersProvider() );
+    CORE_ASSERT( c != nullptr, "Gizmo not set (or with wrong state controler?" );
+    return const_cast<Gizmo::UiSelectionControler*>( c );
+}
+
+Gizmo::UiSelectionControler::UiSelectionControler( std::shared_ptr<Ra::Engine::Material>& material,
+                                                   const Core::Utils::Color& selectedColor ) :
+    ShaderParameterProvider(), m_associatedMaterial( material ), m_selectedColor( selectedColor ) {}
+
+void Gizmo::UiSelectionControler::updateGL() {
+    m_associatedMaterial->updateGL();
+    m_renderParameters = m_associatedMaterial->getParameters();
+    if ( m_selected ) { m_renderParameters.addParameter( "material.color", m_selectedColor ); }
+}
+
+void Gizmo::UiSelectionControler::toggleState() {
+    m_selected = !m_selected;
+    if ( m_selected ) { m_renderParameters.addParameter( "material.color", m_selectedColor ); }
+    else
+    {
+        // Todo : find a way to remove this dynamic cast
+        auto m = dynamic_cast<Engine::PlainMaterial*>( m_associatedMaterial.get() );
+        CORE_ASSERT( m != nullptr, "UiSelectionControler only designed for Plain Material." );
+        m_renderParameters.addParameter( "material.color", m->m_color );
+    }
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Either the condition 'm!=nullptr' is redundant or there is possible null pointer dereference: m.](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830)